jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java

```diff
@@ -148,13 +150,45 @@ public static BigInteger modInverse(BigInteger val, BigInteger modulus) {
   }
 
   /**
-   * Calculate legendre symbol a|p.
+   * Calculate jacobi symbol a|p.
    *
    * @param a must be positive
    * @param p the modulus must be odd
    * @return a|p
    * @throws IllegalArgumentException if a is not positive, or p is not positive, or p is not odd
    */
+  public static int jacobi(BigInteger a, BigInteger p) {
+    if (a.signum() < 0) {
+      throw new IllegalArgumentException("a must be non-negative");
+    }
+    if (p.signum() < 0) {
+      throw new IllegalArgumentException("p must be non-negative");
+    }
+    if (!p.testBit(0)) {
+      throw new IllegalArgumentException("p must be odd");
+    }
+    return INSTANCE.get().jacobiImpl(a, p);
+  }
+
+  /**
+   * Calculate kronecker symbol a|p.
+   *
+   * @param a an integer
+   * @param p the modulus
+   * @return a|p
+   */
+  public static int kronecker(BigInteger a, long p) {
```

Comment:
I think we should just use mpz_kronecker (const mpz_t A, const mpz_t B) to allow both arguments to be BigInteger.  Do you have an compelling reason otherwise?  In generally we've opted for this approach instead of trying to target precision-specific functions.